vm/natives.c

```diff
@@ -1138,5 +1138,393 @@ const AsebaNativeFunctionDescription AsebaNativeDescription_rand =
+	uint16 index_val = vm->variables[index];
+	
+	// variable size
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+	
+	// Check for exception
+	if (vm->variables[index] < 0 || index_val > dest_length - 2 - 1
+		|| src_length > dest_length - 2 - vm->variables[dest])
+	{
+		vm->flags = ASEBA_VM_STEP_BY_STEP_MASK;
+		AsebaSendMessage(vm, ASEBA_MESSAGE_ARRAY_ACCESS_OUT_OF_BOUNDS, &(vm->pc), sizeof(vm->pc));
+		return;
+	}
+	
+	// Do it
+	_AsebaNative_deqinsert(vm, dest, src, index_val, dest_length, src_length);
+}
+
+const AsebaNativeFunctionDescription AsebaNativeDescription_deqinsert =
+{
+	"deque.insert",
+	"shift dest at index by len(src) and insert src",
+	{
+		{ -1, "dest" },
+		{ -2, "src" },
+		{  1, "index" },
+		{ 0, 0 }
+	}
+};
+	
+static void _AsebaNative_deqerase(AsebaVMState *vm, uint16 dest, uint16 index_val, uint16 len_val, uint16 dest_length)
+{
+	// infer deque parameters
+	uint16 dq_size = vm->variables[dest];
+	uint16 dq_start = vm->variables[dest + 1];
+	uint16 dq_capacity = dest_length - 2;
+	
+	// Erase elements as a block
+	// if in left half, shift prefix elements right
+	if (index_val < dq_size / 2)
+	{
+		deque_shift(vm, dest, dq_capacity,
+					/* from */ dq_start + index_val + len_val - 1,
+					/* to   */ dq_start + len_val,
+					-len_val);
+		vm->variables[dest + 1] = dq_start = (dq_start + len_val) % dq_capacity;
+	}
+	// else in right half, shift suffix elements left
+	else
+	{
+		deque_shift(vm, dest, dq_capacity,
+					/* from */ dq_start + index_val,
+					/* to   */ dq_start + dq_size - 1 - len_val,
+					len_val);
+	}
+	vm->variables[dest] = dq_size = dq_size - len_val;
+}
+
+void AsebaNative_deqerase(AsebaVMState *vm)
+{
+	// variable pos
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 index = AsebaNativePopArg(vm);
+	uint16 len = AsebaNativePopArg(vm);
+	uint16 index_val = vm->variables[index];
+	uint16 len_val = vm->variables[len];
+	
+	// variable size
+	uint16 dest_length = AsebaNativePopArg(vm);
+	
+	// Check for exception
+	if (vm->variables[index] < 0 || index_val > dest_length - 2 - 1
+		|| len_val > vm->variables[dest] - index_val)
+	{
+		vm->flags = ASEBA_VM_STEP_BY_STEP_MASK;
+		AsebaSendMessage(vm, ASEBA_MESSAGE_ARRAY_ACCESS_OUT_OF_BOUNDS, &(vm->pc), sizeof(vm->pc));
+		return;
+	}
+	
+	// Do it
+	_AsebaNative_deqerase(vm, dest, index_val, len_val, dest_length);
+}
+
+const AsebaNativeFunctionDescription AsebaNativeDescription_deqerase =
+{
+	"deque.erase",
+	"shift dest at index by len to erase",
+	{
+		{ -1, "dest" },
+		{  1, "index" },
+		{  1, "len" },
+		{ 0, 0 }
+	}
+};
+
+void AsebaNative_deqpushfront(AsebaVMState *vm)
+{
+	// Pop arguments off
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 src = AsebaNativePopArg(vm);
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+
+	// Now call insert with proper arguments
+	_AsebaNative_deqinsert(vm, dest, src, 0, dest_length, src_length);
```

Comment:
Actually you were right, the capacity check was only in the façade! Fixed in af131f9fc220c34ae98a498ac978bf3429a38313, which also has error tests for `push` and `pop`.